openss7-0.9.2.D.SLES9_K2.6.5_rc1(1)/openss7-0.9.2.D.SLES9_K2.6.5_rc1/LiS/man/man3/fdetach.3.man
'\" rtp
.\" -*- nroff -*- vim: ft=nroff nocin nosi
.\"
.\" @(#) fdetach.3.man,v 1.1.4.3 2006/04/05 09:42:34 brian Exp
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2001-2006  OpenSS7 Corporation <http://www.openss7.com/>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not
.\" have taken the same level of care in the production of this manual,
.\" which is licensed free of charge, as they might when working
.\" professionally.  The author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software
.\" on behalf of the U.S. Government ("Government"), the following
.\" provisions apply to you.  If the Software is supplied by the Department
.\" of Defense ("DoD"), it is classified as "Commercial Computer Software"
.\" under paragraph 252.227-7014 of the DoD Supplement to the Federal
.\" Acquisition Regulations ("DFARS") (or any successor regulations) and the
.\" Government is acquiring only the license rights granted herein (the
.\" license rights customarily provided to non-Government users).  If the
.\" Software is supplied to any unit or agency of the Government other than
.\" DoD, it is classified as "Restricted Computer Software" and the
.\" Government's rights in the Software are defined in paragraph 52.227-19
.\" of the Federal Acquisition Regulations ("FAR") (or any successor
.\" regulations) or, in the cases of NASA, in paragraph 18.52.227-86 of the
.\" NASA Supplement to the FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.\" Last Modified 2006/04/05 09:42:34 by brian
.\"
.\" -------------------------------------------------------------------------
.\"
.\" fdetach.3.man,v
.\" Revision 1.1.4.3  2006/04/05 09:42:34  brian
.\" - moved .so line and updated headers
.\"
.\" =========================================================================
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database lis.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
.R2
.so lis.macros
.\" =========================================================================
.\"
.\" Copyright (C) 2000  John A. Boyd Jr.  protologos, LLC
.\"
.\" This library is free software; you can redistribute it and/or
.\" modify it under the terms of the GNU Library General Public
.\" License as published by the Free Software Foundation; either
.\" version 2 of the License, or (at your option) any later version.
.\"
.\" This library is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" Library General Public License for more details.
.\"
.\" You should have received a copy of the GNU Library General Public
.\" License along with this library; if not, write to the
.\" Free Software Foundation, Inc., 59 Temple Place - Suite 330, Cambridge,
.\" MA 02139, USA.
.\"
.\" =========================================================================
.\"
.\"
.TH FDETACH 3 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Library Calls"
.SH NAME
.B fdetach
\- detach any \fISTREAMS\fR file from a path name in the file system
.\"
.\"
.SH SYNOPSIS
.PP
.B "#include <stropts.h> "
.HP
.BI "int fdetach( const char *" path " ); "
.\"
.\"
.SH ARGUMENTS
.PP
.TP
.I path
the filesystem path from which to detach any attached streams.
.\"
.\"
.SH DESCRIPTION
.PP
.BR fdetach "() "
detaches an open
.I STREAMS
file from a path name in the file system.
.BR fdetach "() "
will also
.BR close "(2) "
the
.I STREAMS
file if it represents the last
.BR open "(2) "
of that file.
.\"
.\"
.SH "RETURN VALUE"
.PP
Upon successful completion,
.BR fdetach "() "
returns 0.  Upon failure,
.B fattach
return -1, and
.BR errno "(3) "
is set to indicate the error causing the failure (see below).
.\"
.\"
.SH ERRORS
.PP
.BR fdetach "() "
may fail with at least the following error indications:
.TP \w'[TNOTSUPPORT]\(em'u
.B "[EINVAL]"
No
.I STREAMS
file is currently attached to the
.I path
given as an argument.
.\"
.\"
.SH "APPLICATION USAGE"
.PP
Applications wishing to use the version of the
.BR fdetach "() "
function described herein must link with the
.B LiS
.[
LiS
.]
library which
provides it.  The
Linux/GNU C library otherwise provides a stub supporting calls to
.BR fdetach "() "
which only generates an error indicating that it is not implemented.
.\"
.\"
.SH WARNINGS
.PP
The effect of calling
.BR fdetach "() "
on an attached
.I STREAMS
file occurs automatically when a
.I hangup
indication is received, unless passed file descriptors (in the form of
.B M_PASSFP
messages) are pending receipt via the
.B I_RECVFD
.I ioctl.
This will occur normally in two situations: when the file is a
pipe end and the other end is closed for the last time and released;
and when a driver sends an
.B M_HANGUP
message upward to the stream head.
In this state, the
.I STREAMS
file becomes unusable for normal message traffic, and
the automatic detachment ensures that it will be closed and its
resources released.
It should be noted, however, that this behavior is not common to
other
.ISTREAMS
implementations and should be considered a non-standard extension.
.\"
.\"
.SH "FUTURE DIRECTIONS"
.PP
.BR fdetach "() "
is currently implemented as an IOCTL operation on an anonymous
.I STREAMS
file.
Should
.B LiS
.[
LiS
.]
ever become part of the Linux kernel, it may become a
system call; a system call implementation is in fact already
included within
.B LiS\c
.
.[
LiS
.]
.\"
.\"
.SH BUGS
.PP
.BR fdetach ()
has no known bugs.
.\"
.\"
.SH EXAMPLES
.PP
None.
.\"
.\"
.SH "SEE ALSO"
.PP
.BR fattach "(3), "
.BR fattach "(8), "
.BR fdetach "(8), "
.BR STREAMS "(4) "
.\"
.\"
.SH COMPATIBILITY
.PP
.BR fdetach ()
is compatible with SVID,
.[
svid
.]
XID,
.[
xbd5
.]
SUSv2,
.[
susv2
.]
SUSv3,
.[
susv3
.]
POSIX,
.IR Solaris \(rg,
.IR UnixWare \(rg
and LiS.
.[
LiS
.]
.\"with the following exceptions:
.\".IP \(em \w'\(em\(em'u
.\"
.\"
.SH CONFORMANCE
.PP
SVID,
.[
svid
.]
XID,
.[
xbd5
.]
SUSv2,
.[
susv2
.]
SUSv3,
.[
susv3
.]
POSIX.
.\"
.\"
.SH VERSIONS
.PP
This manpage was written for
.B LiS\c
.
.[
LiS
.]
.\"
.\"
.SH HISTORY
.PP
UNIX\(rg\ System V Release 4 (SVR4)
.[
svr4
.]
.\"
.\"
.SH AUTHOR
.PP
John Boyd, protologos LLC.
.[
$LIST$
.]
.T3


openss7-0.9.2.D.SLES9_K2.6.5_rc1(1)/openss7-0.9.2.D.SLES9_K2.6.5_rc1/streams/lib/libc/fdetach.c
/*****************************************************************************

 @(#) fdetach.c,v OpenSS7-0_9_2_D_rc2(0.9.2.7) 2005/07/18 12:06:58

 -----------------------------------------------------------------------------

 Copyright (c) 2001-2005  OpenSS7 Corporation <http://www.openss7.com>
 Copyright (c) 1997-2000  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software; you can redistribute it and/or modify it under
 the terms of the GNU General Public License as published by the Free Software
 Foundation; either version 2 of the License, or (at your option) any later
 version.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 details.

 You should have received a copy of the GNU General Public License along with
 this program; if not, write to the Free Software Foundation, Inc., 675 Mass
 Ave, Cambridge, MA 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 -----------------------------------------------------------------------------

 Last Modified 2005/07/18 12:06:58 by brian

 *****************************************************************************/

#ident "@(#) fdetach.c,v OpenSS7-0_9_2_D_rc2(0.9.2.7) 2005/07/18 12:06:58"

static char const ident[] =
    "fdetach.c,v OpenSS7-0_9_2_D_rc2(0.9.2.7) 2005/07/18 12:06:58";

#define _XOPEN_SOURCE 600
#define _GNU_SOURCE 1
#define _REENTRANT
#define _THREAD_SAFE

#include <errno.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/types.h>
#include <stropts.h>
#include <sys/stat.h>
#include <sys/ioctl.h>

#include <pthread.h>

extern int __pthread_setcanceltype(int type, int *oldtype);

#pragma weak __pthread_setcanceltype
#pragma weak pthread_setcanceltype

int
pthread_setcanceltype(int type, int *oldtype)
{
	if (__pthread_setcanceltype)
		return __pthread_setcanceltype(type, oldtype);
	if (oldtype)
		*oldtype = type;
	return (0);
}

#define DUMMY_STREAM "/dev/fifo.0"	/* FIXME: /dev/stream,... */
#define DUMMY_MODE   O_RDWR|O_NONBLOCK

static int
__fdetach(const char *path)
{
	int ffd, error = 0;

	if ((ffd = open(DUMMY_STREAM, DUMMY_MODE)) < 0)
		return -1;
	if (ioctl(ffd, I_FDETACH, path) < 0)
		error = errno;
	close(ffd);
	if (error) {
		errno = error;
		return -1;
	}
	return 0;
}

/**
 * @ingroup libLiS
 * @brief detach a path from a stream.
 * @param path the path in the filesystem from which to detach.
 *
 * fdetach() cannot contain a thread cancellation point (SUS/XOPEN/POSIX).  We
 * must protect from asyncrhonous cancellation between the open(), ioctl() and
 * close() operations.
 */
int
fdetach(const char *path)
{
	int oldtype, ret;

	pthread_setcanceltype(PTHREAD_CANCEL_DEFERRED, &oldtype);
	ret = __fdetach(path);
	pthread_setcanceltype(oldtype, NULL);
	return (ret);
}
